### PR TITLE
Enabling use of pyrms for RMG Simulations and Edge Analysis

### PR DESCRIPTION
**Pull request by mjohnson541**
This PR integrates pyrms into RMG enabling us to very easily write new reactor types running simulations and Edge Analysis in RMS.  

There's a lot of small details that are worth mentioning: 

Potentially farther reaching changes:
  - Species labels are forced to be unique
  - Kinetics and thermo are generated when the Species/Reaction objects are created instead of later in enlarge
 
As far as I can tell we seem to be unable to mix pyjulia and nosetests. This put some significant constraints on the structure of these changes. I managed to limit all the rms function calls to the new reactors.py file so this has almost no impact on our current testing infrastructure, however it does mean that the infrastructure in the reactors.py file can't be tested using our current nosetests framework. (This also means codecov is not going to like this PR). 

Things this PR does not enable for RMS reactors: 
  - Ranged reactors
  - Pressure dependence
  - Sensitivity analysis
  - save_simulation_profile
  - Surface (core-edge-surface) functionality (plan is to deprecate this since the branching algorithm is both simpler and     more effective)

Since this PR is already very complex I elected to initially avoid some of the more complex RMG integrations, however none of these (apart from the surface one we were already planning to get rid of) will be very hard to do in the near future. 

This PR is dependent on https://github.com/ReactionMechanismGenerator/ReactionMechanismSimulator.jl/pull/149 and will only pass tests once that RMS PR is merged.

### CODE REVIEW

**Comment by mjohnson541** [comment]
Hmm...should we just have a testing file for reactors.py that we just run as a script during testing?

**Comment by rwest** [comment]
Is pyRMS used for all reactor types now, or just for that new constant volume ideal gas?
I ask because we're about to merge (#2090) coverage-dependent kinetics for surface reactions, which would presumably need to be implemented in RMS?

**Comment by mjohnson541** [comment]
The intent long term is to switch over and deprecate the old reactors, but not in this PR. This just sets up the constant volume reactor as an example for testing the PR all the old reactors still run. 

I'm super happy to add coverage dependent kinetics to RMS, although it might be easier to do after you've merged that. Do you guys have documentation for the format somewhere?

**Review by rwest — COMMENTED** [review]
> I'm super happy to add coverage dependent kinetics to RMS, although it might be easier to do after you've merged that. Do you guys have documentation for the format somewhere?

Great!  Intent is to be same as Chemkin and Cantera. We can dig out the links for you (but not here and now)

File: rmgpy/rmg/model.py
Comment (hwpang) [inline rmgpy/rmg/model.py:652]: Not sure how important it is, but maybe we could move this logging info line somewhere for the user

File: rmgpy/rmg/input.py
Comment (hwpang) [inline rmgpy/rmg/input.py:418]: I think we need to convert this to number of moles because RMS assumes the initial conditions are fed in number of moles?

**Review by hwpang — CHANGES_REQUESTED** [review]
Thanks for the PR! I haven't finished reading yet. I have some comments/requests so far.

File: rmgpy/rmg/reactors.py
Comment (mjohnson541) [reply rmgpy/rmg/reactors.py:388]: So it's worth noting that as of Python 3 keys() doesn't return a list it returns an iterable view of the dictionary keys. Also the only source I've found: https://www.geeksforgeeks.org/complexity-cheat-sheet-for-python-operations/ claims the first of those is O(N) not O(1). I've adopted their technique of calling .get(key) which gives None if there isn't a value.

File: rmgpy/rmg/reactors.py
Comment (mjohnson541) [reply rmgpy/rmg/reactors.py:76]: Is this better?
```
If the species is found
removes that species and any reactions it is involved in 
from the PhaseSystem and its Phases/Interfaces
```

Comment (hwpang) [reply rmgpy/rmg/reactors.py:76]: Yes! Thanks!

File: rmgpy/rmg/reactors.py
Comment (hwpang) [reply rmgpy/rmg/reactors.py:388]: Sounds good! Thanks!

File: rmgpy/rmg/model.py
Comment (mjohnson541) [reply rmgpy/rmg/model.py:75]: I've tried to make this make a bit more sense. In reality ReactionModel for now only really works for one phase or two phases with an interface where one of them is a surface phase so those inputs for the ReactionModel during an RMG job are kind of moot until we do a bit more restructuring in ReactionModel.

File: rmgpy/rmg/model.py
Comment (mjohnson541) [reply rmgpy/rmg/model.py:72]: removed

File: rmgpy/solver/base.pyx
Comment (mjohnson541) [reply rmgpy/solver/base.pyx:59]: Did imports explilcitly

File: rmgpy/rmg/reactors.py
Comment (mjohnson541) [reply rmgpy/rmg/reactors.py:336]: Thanks! Fixed!

File: rmgpy/rmg/reactors.py
Comment (mjohnson541) [reply rmgpy/rmg/reactors.py:357]: Yep! Fixed!

File: rmgpy/rmg/reactors.py
Comment (mjohnson541) [reply rmgpy/rmg/reactors.py:161]: added

File: rmgpy/rmg/input.py
Comment (mjohnson541) [reply rmgpy/rmg/input.py:418]: Since they feed in mole fractions the appropriate scale factor would be 1.0. More specifically for a closed homogeneous reactor the extensive size of the phase is just a scale factor so it doesn't matter whether we scale the moles up or down. It is going to matter a lot for open systems and multiphase kinetics though and we need to start thinking about how we let users specify stuff in more complicated reactors.

**Comment by mjohnson541** [comment]
Okay, I think I've fixed everything!

File: rmgpy/rmg/input.py
Comment (hwpang) [inline rmgpy/rmg/input.py:1097]: In what case a reactionSystem won't be a `Reactor` again? Are we keeping the old base.py objects? Also could you change reactionSystem to reaction_system, thanks!

Comment (mjohnson541) [reply rmgpy/rmg/input.py:1097]: Exactly, for now we're keeping the old base.pyx objects which won't be Reactor objects

**Review by hwpang — CHANGES_REQUESTED** [review]
Only a few more changes!

File: rmgpy/rmg/input.py
Comment (mjohnson541) [reply rmgpy/rmg/input.py:1097]: fixed

File: rmgpy/rmg/reactors.py
Comment (mjohnson541) [reply rmgpy/rmg/reactors.py:58]: XD Looks like I added both with the wildcard and without, fixed.

File: rmgpy/rmg/reactors.py
Comment (mjohnson541) [reply rmgpy/rmg/reactors.py:416]: You mean like as a postprocessing thing? If someone wanted to add that in I don't think it would be very problematic to add the adjlist in here which I suspect is the fastest (also it works even if there are surface species), but for the moment if we don't have a use for it I think it's better to avoid spending time generating them and memory storing them.

File: rmgpy/rmg/reactors.py
Comment (mjohnson541) [reply rmgpy/rmg/reactors.py:376]: fixed

File: rmgpy/rmg/reactors.py
Comment (hwpang) [inline rmgpy/rmg/reactors.py:120]: Did this try to initialize for the `phase_label` below?

Comment (mjohnson541) [reply rmgpy/rmg/reactors.py]: Maybe in a prior iteration of the function XD. Removed.

**Comment by rwest** [comment]
Regarding your nosetests comments in the opening message of the PR, perhaps we should open a separate issue. I gather nosetests is no longer maintained and perhaps we should move to [pytest](https://docs.pytest.org/en/latest/) or something else (maybe even [nose2](https://github.com/nose-devs/nose2)?). Not sure if either of those would solve your problems?

File: rmgpy/rmg/reactors.py
Comment (hwpang) [reply rmgpy/rmg/reactors.py:414]: I agree!

**Comment by hwpang** [comment]
@mjohnson541 With the new push, this PR looks good to me and I think it could be merged after the tests passed!

**Review by hwpang — DISMISSED** [review]
@mjohnson541 With the new push, this PR looks good to me and I think it could be merged after the tests passed!

**Comment by hwpang** [comment]
I also agree with @rwest that we should be looking into alternative testing framework to make sure we can test these rms function calls as I expect more and more would be added to RMG.

**Comment by mjohnson541** [comment]
So pytest at least seems to have the same issue. After a bit more experimentation, I do have a way that I believe should work to test these files in nosetests with the caveat that it's super super slow and adds ~5.5 minutes of compiling to the testing process. 

What might be better is if we could get around using python-jl altogether...which I'm pretty sure would solve this issue...and remove some other constraints too...I'm going to poke the pyjulia people and see if the Julia 1.6 new preferences system makes this feasible. 

However, I don't think that particular issue is going to get resolved in the immediate future so I think we should go ahead and merge.

File: documentation/source/users/rmg/installation/anacondaDeveloper.rst
Comment (hwpang) [inline documentation/source/users/rmg/installation/anacondaDeveloper.rst:50]: The asterisk here got interpreted as italic. Could you add a backslash to escape the asterisk?

Comment (hwpang) [reply documentation/source/users/rmg/installation/anacondaDeveloper.rst]: I realize it looks normal in the rich diff, so it's probably fine

Comment (mjohnson541) [reply documentation/source/users/rmg/installation/anacondaDeveloper.rst]: Sorry, I fixed it a little bit ago, but didn't comment here.

**Review by hwpang — APPROVED** [review]
@mjohnson541 With the newest installation step, I was able to install everything properly and ran the provided example test at `examples/rmg/rms_constant_V/input.py` successfully. I made a couple of small changes for the documentation, one for the formatting, and the other to change the package name from position variable to keyword variable (the former gave errors while I installed). I think we could merge this PR after the tests passes.

**Comment by hwpang** [comment]
@mjohnson541 Could you rebase?

**Review by hwpang — APPROVED** [review]

**Pull request merged**
